src/Test.Utilities/CodeFixRunner.cs

```diff
@@ -0,0 +1,299 @@
+        private readonly CodeFixProvider _codeFixProvider;
+        private readonly TestValidationMode _validationMode;
+        private readonly ISet<string> _fixableDiagnosticIds;
+        private readonly Func<IEnumerable<Diagnostic>, ImmutableArray<Diagnostic>> _getFixableDiagnostics;
+
+        public CodeFixRunner(
+            DiagnosticAnalyzer analyzerOpt,
+            CodeFixProvider codeFixProvider,
+            TestValidationMode validationMode)
+        {
+            _analyzerOpt = analyzerOpt;
+            _codeFixProvider = codeFixProvider;
+            _validationMode = validationMode;
+            _fixableDiagnosticIds = _codeFixProvider.FixableDiagnosticIds.ToSet();
+            _getFixableDiagnostics = diags =>
+                diags.Where(d => _fixableDiagnosticIds.Contains(d.Id)).ToImmutableArrayOrEmpty();
+        }
+
+        public Solution ApplySingleFix(
+            Project project,
+            IEnumerable<TestAdditionalDocument> additionalFiles,
+            int codeFixIndex,
+            int fixableDiagnosticIndex = 0)
+        {
+            var compilation = project.GetCompilationAsync().Result;
+            var analyzerDiagnostics = GetSortedDiagnostics(compilation, additionalFiles: additionalFiles);
+            var compilerDiagnostics = compilation.GetDiagnostics();
+            var fixableDiagnostics = _getFixableDiagnostics(analyzerDiagnostics.Concat(compilerDiagnostics));
+
+            var actions = new List<CodeAction>();
+            var diagnostic = fixableDiagnostics[fixableDiagnosticIndex];
+            var document = FindDocument(diagnostic, project);
+            var context = new CodeFixContext(document, diagnostic, (a, d) => actions.Add(a), CancellationToken.None);
+            _codeFixProvider.RegisterCodeFixesAsync(context).Wait();
+
+            if (!actions.Any())
+            {
+                return project.Solution;
+            }
+
+            if (codeFixIndex >= actions.Count)
+            {
+                throw new Exception($"Unable to invoke code fix at index '{codeFixIndex}', only '{actions.Count}' code fixes were registered.");
+            }
+
+            return DiagnosticFixerTestsExtensions.Apply(actions.ElementAt(codeFixIndex));
+        }
+
+        public Solution ApplyFixAll(
+            Solution solution,
+            IEnumerable<TestAdditionalDocument> additionalFiles,
+            bool allowNewCompilerDiagnostics)
+        {
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                var compilation = project.GetCompilationAsync().Result;
+                var analyzerDiagnostics = GetSortedDiagnostics(compilation, additionalFiles);
+                var compilerDiagnostics = compilation.GetDiagnostics();
+
+                var fixAllProvider = _codeFixProvider.GetFixAllProvider();
+                var diagnosticProvider = new FixAllDiagnosticProvider(this, additionalFiles);
```

Comment:
Can we use the same approach as `ApplySingleFix` to compute `fixableDiagnostics` here and pass it directly to `FixAllDiagnosticProvider` constructor? This should simplify the implementation of `FixAllDiagnosticProvider` which just returns filters its `_fixableDiagnostics` by the requested scope: document, project or solution.

Reply:
Great idea! Thank you!

---
In reply to: [160528829](https://github.com/dotnet/roslyn-analyzers/pull/1514#discussion_r160528829) [](ancestors = 160528829)